### DISCUSSION TITLE
Do open positions/trades factor into the available liquidity/equity calculation?

### DISCUSSION THREAD

**Discussion by kbs-code** [kernc/backtesting.py#1141] [Q&A] [answered]
In the docs there is a part that explains order sizing under Class Order > var size (https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#gsc.tab=0) :

"Order size (negative for short orders).
    
 If size is a value between 0 and 1, it is interpreted as a fraction of current available liquidity (cash plus [Position.pl](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Position.pl) minus used margin). A value greater than or equal to 1 indicates an absolute number of units."

Ok, so if I were to this this code `self.buy(size=.02)` that means "2% of current available liquidity".  And liquidity is "cash plus [Position.pl] minus used margin".  Putting aside the margin part, what does that mean?  To me, it could mean:

1. It's cash (which is a value that never changes after you set initial cash) plus p/l of closed positions.
2. It's cash (which is a value that gets updated after positions close).  
3. It's cash (which is a value that gets updated after positions close) plus p/l of open positions.

Concerns:

My concern at this point is that I do NOT want available liquidity to use my open positions, because I do not trade that way.  When I size my trades, I size it as "2% of my account balance / cash".  If I have open positions, I ignore them when calculating my trade size.  My account balance / cash / availably liquidity should change only when positions are closed.  If this is not the way the backtester works, then I need to modify the code for my own purposes.  Also, the equity / available liquidity may also use margin in its calculation.  If this is the case I need to take that out.  

Also, can someone please clarify how the program defines "positions" and "trades"?  Could a trade be an open position?  Are trades simply positions that have been closed?  For me, positions get added when I place a trade or order.  A position shows it's current p/l.  If it gets closed, it becomes part of my trade history / journal.  Also, in the eyes of the program, what is the difference between "equity", "cash", and "available liquidity"?

To try and understand more clearly, I went to the code in `backtesing.py` and found line 788:

    @property
        def equity(self) -> float:
            return self._cash + sum(trade.pl for trade in self.trades)

Ok so here we see that equity is defined as "cash plus the sum of some p/l figures that I don't know".  This doesn't clarify my three questions so I ask myself, "What is `self.trades`?".  Well I think that is in line 718:

  
    self.trades: List[Trade] = []
   
This has me confused.  Can I see what goes in this list?

There is also this part:
`trade.pl`

I tried to look for this and I think I found it in line 637:

    @property
    def pl(self):
        """Trade profit (positive) or loss (negative) in cash units."""
        price = self.__exit_price or self.__broker.last_price
        return self.__size * (price - self.__entry_price)
        
At this point I'm still wondering whether these trades are closed or open positions, and I don't know where to look in the code to answer that.  So again my main question is: How is equity/liquidity calculated?  If it's cash plus open positions, that's wrong for my use case and I need to figure out a way to change it in the code.  I'm trying to size my positions in terms of a percentage of cash or account balance, which is something that changes after positions close.

**Comment by initial-d** [comment, marked as answer]
My reading of that docs sentence is:

`size=0.02` means "use 2% of current available liquidity for this order", not "risk 2% of account equity" and not necessarily "buy 2% of starting cash".

Available liquidity is roughly:

`cash + current unrealized P/L - margin already used`

So open positions can affect it in two ways:

- their unrealized P/L changes the liquidity/equity available for new fractional-size orders;
- their used margin reduces what is available for additional orders.

But the market value of an existing open position is not simply treated as free cash that can be reused. If you already have a position open and submit another `self.buy(size=0.02)`, Backtesting.py will size that new order against the available liquidity at that moment. With `exclusive_orders=False`, that can lead to multiple overlapping long trades unless you block new entries yourself.

For example, if you want only one long position at a time, gate the entry:

```python
if not self.position.is_long:
    self.buy(size=0.02)
```

Also, if your intention is "risk 2% if the stop-loss is hit", `size=0.02` is not the right calculation. In that case you need to compute units from the stop distance, e.g. account_risk / abs(entry_price - stop_price), then pass an absolute unit size greater than or equal to 1. Fractional `size < 1` is capital allocation/liquidity sizing, not stop-risk sizing.
